metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala

```diff
@@ -2215,16 +2215,24 @@ class MetalsLanguageServer(
   }
 
   private def indexWorkspaceSources(): Unit = {
+    case class SourceToIndex(
+        source: AbsolutePath,
+        sourceItem: AbsolutePath,
+        targets: Iterable[b.BuildTargetIdentifier]
+    )
+    val sourcesToIndex = mutable.ListBuffer.empty[SourceToIndex]
     for {
       (sourceItem, targets) <- buildTargets.sourceItemsToBuildTargets
       source <- sourceItem.listRecursive
       if source.isScalaOrJava
     } {
-      targets.asScala.foreach { target =>
-        buildTargets.linkSourceFile(target, source)
-      }
-      indexSourceFile(source, Some(sourceItem), targets.asScala.headOption)
+      targets.asScala.foreach(buildTargets.linkSourceFile(_, source))
+      sourcesToIndex += SourceToIndex(source, sourceItem, targets.asScala)
     }
+
+    sourcesToIndex.par.foreach(f =>
```

Comment:
@kpodsiad mentioned that this way indexing is using up all the cores of the machine. I wonder whether this will be ok for the users? Is it possible to work on other things in the meantime? Or should we reduce paralelization to half the processors?

Reply:
I've limited it to cores/2 which I think should be ok as I think this should be mostly File IO limited.

@kpodsiad Could you check on your machine again if this is easier on cores without losing performance.  I find looking at core usage on WSL doesn't seem to work very well

Reply:
@Arthurm1 cores/2 case is much lighter in terms of CPU load (no surprise - half of the cores are free) and indexing duration isn't much longer.
1 core -> cores/2-> all cores
kafka: 22s -> 14s -> 12s

I'm ok with using all cores since indexing is taking only a few seconds, but for more conservative users cores/2 will be better.

Reply:
Thanks for testing